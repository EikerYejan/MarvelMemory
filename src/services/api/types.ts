import { Character } from '../../types'

export type FetchCharactersResponse = {
  attributionHTML: string
  data: {
    results: Character[]
  }
}
