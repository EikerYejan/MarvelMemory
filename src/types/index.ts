export type Card = {
  id: string
  name: string
  image: string
  isSelected?: boolean
  isMatched?: boolean
}

export type Character = {
  id: string
  name: string
  thumbnail: {
    extension: string
    path: string
  }
}
