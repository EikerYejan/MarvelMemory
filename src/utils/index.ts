import { DefaultTheme, StyledProps } from 'styled-components'
import { Card, Character } from '../types'

export const shuffleArray = <T extends Array<unknown>>(array: T) => {
  const shuffled = array

  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = array[i]

    shuffled[i] = array[j]
    shuffled[j] = temp
  }

  return shuffled
}

export const replicateArray = <T extends Array<unknown>>(arr: T, times = 2) => {
  const arrays = Array(times).fill(0)
  const replicated = arrays.map(() => arr)

  return replicated.flat()
}

export const uuid = () => ((1 + Math.random()) * 0x10000 ?? 0).toString(16).substring(1)

export const transformCharacters = (characters: Character[]): Card[] => {
  return characters.map(({ name, thumbnail }) => ({
    id: uuid(),
    name,
    image: `${thumbnail.path}/standard_xlarge.${thumbnail.extension}`,
  }))
}

export const themeColor = (color: keyof DefaultTheme['colors']) => {
  return ({ theme }: StyledProps<DefaultTheme>) => {
    return theme.colors?.[color]
  }
}
