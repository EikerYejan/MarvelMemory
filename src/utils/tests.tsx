import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../theme/defaultTheme'

export const renderWithTheme = (component: React.ReactElement) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
