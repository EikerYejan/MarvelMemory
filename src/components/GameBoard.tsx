import React, { useState, useEffect, useRef } from 'react'
import styled, { keyframes } from 'styled-components'
import { Fireworks } from 'fireworks-js/dist/react'
import Card from './Card'
import { replicateArray, shuffleArray, transformCharacters } from '../utils'
import { fetchCharacters } from '../services/api'
import { Card as CardType } from '../types'

type Props = {
  cardsCount: number
}

const spin = keyframes`
  0% {
    transform: rotate(0deg)
  }

  50% {
    transform: rotate(180deg)
  }

  100% {
    transform: rotate(360deg)
  }
`

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  min-height: 100vh;
  padding: 70px 45px;

  .fireworks {
    position: absolute;
    z-index: 1;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;

    canvas {
      width: 100%;
      height: 100%;
    }
  }

  .grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, 200px);
    place-content: center;
    gap: 20px;
    max-width: 1200px;
    width: 100vw;
  }

  .loader {
    position: absolute;
    width: 45px;
    height: 45px;
    border: 3px solid transparent;
    top: calc(50% - 22.5px);
    left: calc(50% - 22.5px);
    border-left-color: ${({ theme }) => theme.colors.red};
    border-top-color: ${({ theme }) => theme.colors.red};
    border-radius: 50%;
    animation: ${spin} 0.5s infinite linear;
  }

  footer {
    a {
      color: ${({ theme }) => theme.colors.white};
      font-weight: 100;
      margin-top: 55px;
      text-decoration: none;
      display: block;
      text-align: center;
    }
  }
`

const Controls = styled.div`
  margin: 0 0 45px 0;
  display: flex;
  flex-direction: column;
  align-items: flex-end;

  .buttons {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    margin-top: 15px;
  }

  .reset-button {
    background: none;
    border: none;
    padding: 0;
    color: ${({ theme }) => theme.colors.white};
    display: flex;
    align-items: center;
    justify-content: center;

    img {
      filter: invert(1);
      width: 25px;
      height: 25px;
      display: block;
    }
  }

  .cards-count {
    margin-right: 15px;
  }
`

const verifyCards = ([card1, card2]: CardType[]) => {
  return card1.name === card2.name
}

const GameBoard = ({ cardsCount = 20 }: Props) => {
  const [cards, setCards] = useState<CardType[]>([])
  const [selectedCards, setSelectedCards] = useState<CardType[]>([])
  const [matchedCards, setMatchedCards] = useState<CardType[]>()
  const [isCompleted, setIsCompleted] = useState(false)

  const [level, setLevel] = useState(1)
  const [isLoading, setIsLoading] = useState(true)

  const copyrightText = useRef<string>()

  const selectedIds = selectedCards?.map((card) => card.id)
  const selectedCardsCount = selectedCards?.length ?? 0
  const matchedIds = matchedCards?.map((card) => card.id)

  const onSelect = (card: CardType) => {
    if (selectedCardsCount <= 1 && !selectedIds.includes(card.id)) setSelectedCards((prev) => [...(prev ?? []), card])
  }

  const getPageSettings = () => {
    const pageSize = cardsCount / 2
    const skip = pageSize * level

    return [pageSize, skip]
  }

  useEffect(() => {
    const fetchData = async () => {
      try {
        setSelectedCards([])
        setMatchedCards(undefined)
        setIsLoading(true)

        const { characters, attributionText } = await fetchCharacters(...getPageSettings())
        const processedResults = transformCharacters(replicateArray(characters))

        copyrightText.current = attributionText
        setCards(processedResults)
      } catch (error) {
        alert('error')
      } finally {
        setIsLoading(false)
      }
    }

    fetchData()
  }, [cardsCount, level])

  useEffect(() => {
    if (selectedCardsCount >= 2) {
      if (verifyCards(selectedCards)) {
        setTimeout(() => {
          setMatchedCards((prev) => [...(prev ?? []), ...selectedCards])
        }, 500)
      }

      setTimeout(() => setSelectedCards([]), 700)
    }
  }, [selectedCardsCount, selectedCards])

  useEffect(() => {
    if (typeof matchedCards === `object` && matchedCards.length === cards.length) {
      setIsCompleted(true)

      setTimeout(onFinishLevel, 3500)
    }
  }, [matchedCards, cards])

  const onReset = () => {
    setSelectedCards([])
    setMatchedCards(undefined)
    setCards(shuffleArray(cards))
  }

  const onFinishLevel = () => {
    // eslint-disable-next-line
    const moveToNextLevel = confirm('Go to the next level?')

    if (moveToNextLevel) {
      setLevel((prev) => (prev ?? 0) + 1)
      setIsCompleted(false)
    }
  }

  return (
    <Wrapper className="GameBoard">
      {isLoading ? (
        <span className="loader" />
      ) : (
        <div>
          {isCompleted && (
            <div className="fireworks">
              <Fireworks options={{ particles: 100 }} />
            </div>
          )}
          <Controls>
            <p className="cards-count">{cards.length - (matchedCards?.length ?? 0)} cards remainig</p>
            <p className="current-level">Level {level}</p>
            <div className="buttons">
              <button className="reset-button" onClick={onReset} type="button" title="Reset">
                <img src="images/refresh.svg" alt="refresh" />
              </button>
            </div>
          </Controls>
          <div className="grid">
            {cards.map((card) => (
              <Card
                isMatched={matchedIds?.includes(card.id)}
                isSelected={selectedIds?.includes(card.id)}
                onClick={() => onSelect(card)}
                key={card.id}
                {...card}
              />
            ))}
          </div>
          <footer>
            <a title="Marvel website" href="https://marvel.com" target="_blank" rel="noopener noreferrer">
              {copyrightText.current}
            </a>
          </footer>
        </div>
      )}
    </Wrapper>
  )
}

export default GameBoard
