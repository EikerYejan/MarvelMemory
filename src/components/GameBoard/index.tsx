import React, { useState, useEffect, useRef, useMemo } from 'react'
import styled, { keyframes } from 'styled-components'
import { Fireworks } from 'fireworks-js/dist/react'
import Card from '../Card'
import { replicateArray, shuffleArray, transformCharacters } from '../../utils'
import { fetchCharacters } from '../../services/api'
import { Card as CardType } from '../../types'
import ErrorIndicator from '../Error'
import Alert from '../Alert'

type Props = {
  cardsCount: number
  onChangeCardsCount: (count: number) => void
}

const spin = keyframes`
  0% {
    transform: rotate(0deg)
  }

  50% {
    transform: rotate(180deg)
  }

  100% {
    transform: rotate(360deg)
  }
`

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  min-height: 100vh;
  padding: 70px 45px;

  .fireworks {
    position: fixed;
    z-index: 1;
    width: 100vw;
    height: 100vh;
    top: 0;
    left: 0;

    canvas {
      width: 100%;
      height: 100%;
    }
  }

  .grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, 200px);
    place-content: center;
    gap: 20px;
    max-width: 1200px;
    width: 100vw;
  }

  .loader {
    position: absolute;
    width: 45px;
    height: 45px;
    border: 3px solid transparent;
    top: calc(50% - 22.5px);
    left: calc(50% - 22.5px);
    border-left-color: ${({ theme }) => theme.colors.red};
    border-top-color: ${({ theme }) => theme.colors.red};
    border-radius: 50%;
    animation: ${spin} 0.5s infinite linear;
  }

  footer {
    a {
      color: ${({ theme }) => theme.colors.white};
      font-weight: 100;
      margin-top: 55px;
      text-decoration: none;
      display: block;
      text-align: center;
    }
  }
`

const Controls = styled.div`
  margin: 0 auto 45px;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  padding: 0 50px;

  .buttons {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    margin-top: 15px;
  }

  .reset-button,
  .change-cards-count-button {
    background: none;
    border: none;
    padding: 0;
    color: ${({ theme }) => theme.colors.white};
    display: flex;
    align-items: center;
    justify-content: center;

    img {
      filter: invert(1);
      width: 25px;
      height: 25px;
      display: block;
    }

    &:disabled {
      opacity: 0.5;
      pointer-events: none;
    }
  }

  .change-cards-count-button {
    margin-right: 10px;
  }
`

const verifyCards = ([card1, card2]: CardType[]) => {
  return card1.name === card2.name
}

const GameBoard = ({ cardsCount, onChangeCardsCount }: Props) => {
  const [cards, setCards] = useState<CardType[]>([])
  const [selectedCards, setSelectedCards] = useState<CardType[]>([])
  const [matchedCards, setMatchedCards] = useState<CardType[]>()
  const [showConfirmationModal, setShowConfirmationModal] = useState(false)
  const [hasError, setHasError] = useState(false)

  const [level, setLevel] = useState(1)
  const [isLoading, setIsLoading] = useState(true)

  const copyrightText = useRef<string>()

  const selectedIds = selectedCards?.map((card) => card.id)
  const selectedCardsCount = selectedCards?.length ?? 0
  const matchedIds = matchedCards?.map((card) => card.id)

  const onSelect = (card: CardType) => {
    if (selectedCardsCount <= 1 && !selectedIds.includes(card.id)) setSelectedCards((prev) => [...prev, card])
  }

  const pageSettings = useMemo(() => {
    const pageSize = cardsCount / 2
    const skip = pageSize * level

    return [pageSize, skip]
  }, [cardsCount, level])

  const fetchData = async () => {
    try {
      setSelectedCards([])
      setMatchedCards(undefined)
      setHasError(false)
      setIsLoading(true)

      const { characters, attributionText } = await fetchCharacters(...pageSettings)
      const processedResults = shuffleArray(transformCharacters(replicateArray(characters)))

      copyrightText.current = attributionText
      setCards(processedResults)
    } catch (error) {
      setHasError(true)
    } finally {
      setIsLoading(false)
    }
  }

  useEffect(() => {
    fetchData()
  }, [cardsCount, level])

  useEffect(() => {
    if (selectedCardsCount >= 2) {
      if (verifyCards(selectedCards)) {
        setTimeout(() => {
          setMatchedCards((prev) => [...(prev ?? []), ...selectedCards])
        }, 500)
      }

      setTimeout(() => setSelectedCards([]), 700)
    }
  }, [selectedCardsCount, selectedCards])

  useEffect(() => {
    if (typeof matchedCards === `object` && matchedCards.length === cards.length) {
      setTimeout(setShowConfirmationModal, 500, true)
    }
  }, [matchedCards, cards])

  const onReset = () => {
    setSelectedCards([])
    setMatchedCards(undefined)
    setCards(shuffleArray(cards))
  }

  const onFinishLevel = () => {
    setLevel((prev) => prev + 1)
    setShowConfirmationModal(false)
  }

  const renderContent = () => {
    if (isLoading && !hasError) return <span className="loader" />
    if (hasError) return <ErrorIndicator onRetry={fetchData} />

    return (
      <section data-testid="game-board">
        {showConfirmationModal && (
          <div className="fireworks">
            <Alert
              title="Congrats, You won!"
              subtitle="Do you want to move to the next level?"
              cancelButtonText="Nope"
              confirmButtonText="Hell yeah!"
              onCancel={() => setShowConfirmationModal(false)}
              onConfirm={onFinishLevel}
            />
            <Fireworks options={{ particles: 100 }} />
          </div>
        )}
        <Controls>
          <p className="cards-count">{cards.length - (matchedCards?.length ?? 0)} cards remainig</p>
          <p className="current-level">Level {level}</p>
          <div className="buttons">
            <button
              className="change-cards-count-button"
              onClick={() => onChangeCardsCount(0)}
              type="button"
              title="Change cards count"
            >
              <img src="images/dice.svg" alt="refresh" />
            </button>
            <button
              disabled={!matchedCards?.length}
              className="reset-button"
              onClick={onReset}
              type="button"
              title="Reset"
            >
              <img src="images/refresh.svg" alt="refresh" />
            </button>
          </div>
        </Controls>
        <div className="grid">
          {cards.map((card) => (
            <Card
              isMatched={matchedIds?.includes(card.id)}
              isSelected={selectedIds?.includes(card.id)}
              onClick={() => onSelect(card)}
              key={card.id}
              {...card}
            />
          ))}
        </div>
        <footer>
          <a title="Marvel website" href="https://marvel.com" target="_blank" rel="noopener noreferrer">
            {copyrightText.current}
          </a>
        </footer>
      </section>
    )
  }

  return <Wrapper className="GameBoard">{renderContent()}</Wrapper>
}

export default GameBoard
