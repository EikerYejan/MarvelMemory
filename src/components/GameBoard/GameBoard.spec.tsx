import React from 'react'
import axios from 'axios'
import { waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'
import GameBoard from '.'
import { renderWithTheme } from '../../utils/tests'
import { Character } from '../../types'

const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

const characters: Character[] = [
  {
    id: '1',
    name: 'Tony Stark',
    thumbnail: {
      extension: 'jpg',
      path: 'cdn.com/image',
    },
  },
]

const cardsCount = characters.length * 2
const onChangeCardsCount = jest.fn()

jest.mock('axios', () => ({
  __esModule: true,
  default: {
    get: () =>
      Promise.resolve({
        data: {
          attributionHTML: 'HTML',
          attributionText: 'TEXT',
          data: {
            results: characters,
          },
        },
      }),
  },
}))

jest.mock('fireworks-js/dist/react', () => ({
  __esModule: true,
  Fireworks: () => <div>Fireworks</div>,
}))

const finishGame = async () => {
  const { baseElement, getAllByTestId, getByTestId } = renderWithTheme(
    <GameBoard cardsCount={cardsCount} onChangeCardsCount={onChangeCardsCount} />,
  )
  const cards = await waitFor(() => getAllByTestId('card'))
  cards.forEach((card) => {
    userEvent.click(card)
  })

  await act(async () => {
    await timeout(1000)
  })

  return { baseElement, getByTestId, getAllByTestId }
}

describe('<GameBoard />', () => {
  it('should render', async () => {
    const { baseElement, getAllByTestId } = renderWithTheme(
      <GameBoard cardsCount={cardsCount} onChangeCardsCount={onChangeCardsCount} />,
    )
    const cards = await waitFor(() => getAllByTestId('card'))

    expect(baseElement).toBeInTheDocument()
    expect(cards.length).toBe(cardsCount)
  })

  it('Should finish game', async () => {
    const { baseElement, getByTestId } = await finishGame()
    const successAlert = await waitFor(() => getByTestId('alert'))

    expect(baseElement).toMatchSnapshot()
    expect(successAlert).toBeInTheDocument()
  })

  it('Should finish game and move to the next level', async () => {
    const { getByTestId, baseElement } = await finishGame()

    const successAlert = await waitFor(() => getByTestId('alert'))
    const confirmButton = await waitFor(() => successAlert.querySelector<HTMLButtonElement>('button.confirm'))

    expect(confirmButton).toBeInTheDocument()

    await act(async () => {
      confirmButton?.click()

      await timeout(1000)
    })

    expect(baseElement.querySelector('.current-level')).toHaveTextContent('Level 2')
  })

  it('Should finish game and stay in the same level', async () => {
    const { getByTestId, baseElement } = await finishGame()

    const successAlert = await waitFor(() => getByTestId('alert'))
    const cancelButton = await waitFor(() => successAlert.querySelector<HTMLButtonElement>('button.cancel'))

    expect(cancelButton).toBeInTheDocument()

    await act(async () => {
      cancelButton?.click()

      await timeout(1000)
    })

    expect(baseElement.querySelector('.current-level')).toHaveTextContent('Level 1')
  })

  it('Should render Error component', async () => {
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.reject())

    const { getByTestId } = renderWithTheme(
      <GameBoard cardsCount={cardsCount} onChangeCardsCount={onChangeCardsCount} />,
    )
    const errorComponent = await waitFor(() => getByTestId('error'))

    expect(errorComponent).toBeInTheDocument()
  })
})
