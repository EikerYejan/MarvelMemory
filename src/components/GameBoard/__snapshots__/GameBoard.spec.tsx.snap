// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameBoard /> Should finish game 1`] = `
<body>
  <div>
    <div
      class="sc-hKFxyN eTgajs GameBoard"
    >
      <section
        data-testid="game-board"
      >
        <div
          class="fireworks"
        >
          <div
            class="sc-dlnjwi bUtLUq"
            data-testid="alert"
          >
            <button
              class="close"
              title="Close"
              type="button"
            >
              Ã—
            </button>
            <h4>
              Congrats, You won!
            </h4>
            <p>
              Do you want to move to the next level?
            </p>
            <div
              class="buttons"
            >
              <button
                class="cancel"
                title="Close"
                type="button"
              >
                Nope
              </button>
              <button
                class="confirm"
                title="Confirm"
                type="button"
              >
                Hell yeah!
              </button>
            </div>
          </div>
          <div>
            Fireworks
          </div>
        </div>
        <div
          class="sc-eCApnc JSJxT"
        >
          <p
            class="cards-count"
          >
            0
             cards remainig
          </p>
          <p
            class="current-level"
          >
            Level 
            1
          </p>
          <div
            class="buttons"
          >
            <button
              class="change-cards-count-button"
              title="Change cards count"
              type="button"
            >
              <img
                alt="refresh"
                src="images/dice.svg"
              />
            </button>
            <button
              class="reset-button"
              title="Reset"
              type="button"
            >
              <img
                alt="refresh"
                src="images/refresh.svg"
              />
            </button>
          </div>
        </div>
        <div
          class="grid"
        >
          <div
            class="sc-bdnxRM hBUBjL"
            data-testid="card"
            role="button"
            tabindex="0"
          >
            <div
              class="inner"
            >
              <div
                class="front"
              >
                <img
                  alt="marvel-logo"
                  src="https://i.pinimg.com/originals/24/92/00/249200c431fe811110761709b303fcaf.jpg"
                />
              </div>
              <div
                class="back"
              >
                <img
                  alt="Tony Stark"
                  src="cdn.com/image/standard_xlarge.jpg"
                />
                <p>
                  Tony Stark
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-bdnxRM hBUBjL"
            data-testid="card"
            role="button"
            tabindex="0"
          >
            <div
              class="inner"
            >
              <div
                class="front"
              >
                <img
                  alt="marvel-logo"
                  src="https://i.pinimg.com/originals/24/92/00/249200c431fe811110761709b303fcaf.jpg"
                />
              </div>
              <div
                class="back"
              >
                <img
                  alt="Tony Stark"
                  src="cdn.com/image/standard_xlarge.jpg"
                />
                <p>
                  Tony Stark
                </p>
              </div>
            </div>
          </div>
        </div>
        <footer>
          <a
            href="https://marvel.com"
            rel="noopener noreferrer"
            target="_blank"
            title="Marvel website"
          >
            TEXT
          </a>
        </footer>
      </section>
    </div>
  </div>
</body>
`;
